caffe2/python/operator_test/basic_rnn_test.py

```diff
@@ -0,0 +1,139 @@
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+from caffe2.python import workspace, core, scope, rnn_cell
+from caffe2.python.model_helper import ModelHelper
+from caffe2.python.rnn.rnn_cell_test_util import sigmoid, tanh, _prepare_rnn
+import caffe2.python.hypothesis_test_util as hu
+from caffe2.proto import caffe2_pb2
+
+from functools import partial
+from hypothesis import given
+from hypothesis import settings as ht_settings
+import hypothesis.strategies as st
+import numpy as np
+import unittest
+
+def basic_rnn_reference(input, hidden_initial,
+                        i2h_w, i2h_b,
+                        gate_w, gate_b,
+                        seq_lengths,
+                        drop_states,
+                        use_sequence_lengths):
+    D = hidden_initial.shape[-1]
+    T = input.shape[0]
+    N = input.shape[1]
+
+    if seq_lengths is not None:
+        seq_lengths = (np.ones(shape=(N, D)) *
+                       seq_lengths.reshape(N, 1)).astype(np.int32)
+
+    ret = []
+
+    hidden_prev = hidden_initial
+
+    for t in range(T):
+        input_fc = np.dot(   input[t],  i2h_w.T) +  i2h_b
+        recur_fc = np.dot(hidden_prev, gate_w.T) + gate_b
+        hidden_t = tanh(input_fc + recur_fc)
+
+        if seq_lengths is not None:
+            valid = (t < seq_lengths).astype(np.int32)
+            assert valid.shape == (N, D), (valid.shape, (N, D))
+            hidden_t = hidden_t * valid + \
+                       hidden_prev * (1 - valid) * (1 - drop_states)
+
+        ret.append(hidden_t)
+        hidden_prev = hidden_t
+    return ret
+
+
+class BasicRNNCellTest(hu.HypothesisTestCase):
+    @given(
+        seed=st.integers(0, 2**32 - 1),
+        seq_length=st.integers(min_value=1, max_value=5),
+        batch_size=st.integers(min_value=1, max_value=5),
+        input_size=st.integers(min_value=1, max_value=5),
+        hidden_size=st.integers(min_value=1, max_value=5),
+        drop_states=st.booleans(),
+        sequence_lengths=st.booleans(),
+        **hu.gcs
+    )
+    @ht_settings(max_examples=15)
+    def test_basic_rnn(self, seed, seq_length, batch_size, input_size, hidden_size,
+                       drop_states, sequence_lengths, gc, dc):
+        np.random.seed(seed)
+
+        seq_lengths_data = np.random.randint(1, seq_length + 1, size=(batch_size,)).astype(np.int32)
+        input_blob_data = np.random.randn(seq_length, batch_size, input_size).astype(np.float32)
+        initial_h_data = np.random.randn(batch_size, hidden_size).astype(np.float32)
+        gates_t_w_data = np.random.randn(hidden_size, hidden_size).astype(np.float32)
+        gates_t_b_data = np.random.randn(hidden_size).astype(np.float32)
+        i2h_w_data = np.random.randn(hidden_size, input_size).astype(np.float32)
+        i2h_b_data = np.random.randn(hidden_size).astype(np.float32)
+
+        with core.DeviceScope(gc):
```

Comment:
could you use assertReferenceChecks? It would free you from doing temp workspace and manual blob feeding. 

This is a common abstraction we normally use for testing operators in caffe2 when using hypothesis. There are several rnn specific examples in rnn_cell_test.

Reply:
I can't currently, because assertReferencChecks only works with a single operator, it doesn't know how to compare a multi-op chunk of a graph

Reply:
Did you check rnn_cell_test? It handles this by looking into net.op[-1] for most of the RNNCell types.

Reply:
yes, I did try. The problem is that because I don't have a single C++ op, net.op[-1] is not BasicRNNCell, it's something like one of the constituent FCs

Reply:
I don't understand this actually. One of the net's ops has to be RecurrentNetwork operator. Which will have the whole cell inside it. If this is not -1 in your case it can be just another index. You don't get access into the cell itself by navigating net.op